code/modules/mob/living/carbon/xenomorph/castes/pyrogen/abilities_pyrogen.dm

```diff
@@ -0,0 +1,529 @@
+		for(var/mob/living/carbon/fired in affecting)
+			fire_damage = PYROGEN_FIREBALL_AOE_DAMAGE
+			if(fired.has_status_effect(STATUS_EFFECT_MELTING_FIRE))
+				var/datum/status_effect/stacking/melting_fire/debuff = fired.has_status_effect(STATUS_EFFECT_MELTING_FIRE)
+				debuff.add_stacks(PYROGEN_FIREBALL_MELTING_STACKS)
+			fired.apply_status_effect(STATUS_EFFECT_MELTING_FIRE, PYROGEN_FIREBALL_MELTING_STACKS)
+			fired.take_overall_damage(fire_damage, BURN, FIRE, FALSE, FALSE, TRUE, 0)
+	new /obj/effect/temp_visual/xeno_fireball_explosion(get_turf(proj))
+
+
+// ***************************************
+// *********** Firenado
+// ***************************************
+/datum/action/ability/activable/xeno/firestorm
+	name = "Fire Storm"
+	action_icon_state = "whirlwind"
+	desc = "Unleash 3 fiery tornados. They will try to close on your target tile"
+	target_flags = ABILITY_TURF_TARGET
+	ability_cost = 50
+	cooldown_duration = 8 SECONDS
+	keybinding_signals = list(
+		KEYBINDING_NORMAL = COMSIG_XENOABILITY_FIRENADO,
+	)
+
+/obj/effect/xenomorph/firenado
+	name = "Plasma Whirlwind"
+	desc = "A glowing whirlwind of... cold plasma? seems to \"burn\" "
+	icon = 'icons/effects/64x64.dmi'
+	icon_state = "whirlwind"
+	anchored = TRUE
+	mouse_opacity = MOUSE_OPACITY_TRANSPARENT
+	pixel_x = -16
+	/// How many process ticks can we do before we are deleted/we extinguish
+	var/burn_ticks_left = 20
+	/// Target turf to bias going towards
+	var/turf/target
+
+/obj/effect/xenomorph/firenado/New(loc, arg_target)
+	. = ..()
+	target = arg_target
+
+/obj/effect/xenomorph/firenado/Initialize(mapload)
+	. = ..()
+	START_PROCESSING(SSfastprocess, src)
+	var/static/list/connections = list(
+		COMSIG_ATOM_ENTERED = PROC_REF(on_cross),
+	)
+	AddElement(/datum/element/connect_loc, connections)
+
+
+/obj/effect/xenomorph/firenado/Destroy()
+	STOP_PROCESSING(SSfastprocess, src)
+	return ..()
+
+/// mob affecting
+/obj/effect/xenomorph/firenado/proc/mob_act(mob/living/carbon/human/target)
+	var/damage = PYROGEN_TORNADE_HIT_DAMAGE
+	if(target.has_status_effect(STATUS_EFFECT_MELTING_FIRE))
+		var/datum/status_effect/stacking/melting_fire/debuff = target.has_status_effect(STATUS_EFFECT_MELTING_FIRE)
+		debuff.add_stacks(PYROGEN_TORNADO_MELTING_FIRE_STACKS)
+	target.apply_status_effect(STATUS_EFFECT_MELTING_FIRE, PYROGEN_TORNADO_MELTING_FIRE_STACKS)
+	target.take_overall_damage(damage, BURN, FIRE)
+
+///Effects applied to a mob that crosses a burning turf
+/obj/effect/xenomorph/firenado/proc/on_cross(datum/source, mob/living/carbon/human/target, oldloc, oldlocs)
+	if(!istype(target))
+		return
+	mob_act(target)
+	qdel(src)
+
+/obj/effect/xenomorph/firenado/Bump(atom/target)
+	. = ..()
+	if(iswallturf(target))
+		if(!istype(target,/turf/closed/wall/resin))
+			var/turf/closed/wall/wall = target
+			wall.take_damage(PYROGEN_TORNADE_HIT_DAMAGE, BURN)
+			qdel(src)
+			return
+	else if(isobj(target) && !istype(target, type))
+		var/obj/object = target
```

Comment:
tornadoes are all nondense so this second istype is unneeded

Reply:
wheter or not dense , they still managed to bump onto eachother on local testing.

Reply:
```suggestion
	else if(isobj(target) && !istype(target, /obj/effect/xenomorph/firenado))
		var/obj/object = target
```